metabase/metabase
# Add drag and drop reordering of pinned items

Related to #19758.
Resolves #19997.

This PR adds drag-and-drop sorting to pinned collection items. Questions, cards, and models are still split into separate lists, but you can click, drag, and drop within a section to move a given item either forward or backward to reorder the section.

This code uses `react-dnd` as all Collections code does. I tried using `react-sortable-hoc` to do this since parameter widget sorting is so much more pleasant, but `react-dnd` and `react-sortable-hoc` don't really play well together, and we are trying to avoid a separate "edit" mode right now.

I added a few new props to the existing `PinDropTarget` component,`itemModel`, `isFrontTarget`, and `isBackTarget`, all of which are used in the component's `canDrop` function to determine whether or not the drop target component is usable by whatever item you're dragging. `itemModel` is simple; you can't move a dashboard into the question section, etc. 

Concerning `isFrontTarget` and `isBackTarget`, you'll see in the `PinnedItemOverview` that I've added **two** `PinnedItemSortDropTarget` components next to each item component, one in front and one behind. This is so that when you are dragging an item we can add a little bit of styling to show the user where the item will be places upon dropping it.

**Testing**
1. Create a collection and pin questions, dashboards, models.
2. Click and drag pinned items and drop them in drop targets that appear on the page, either in front of or behind the dragged item.
3. Dropping an item should reorder the list, pushing other items back if you move the item forward or pushing other items forward if you move the item backward.

https://user-images.githubusercontent.com/13057258/151447249-3a6ae92a-8b43-4822-9e03-db38c323e267.mov
